src/Spice86.Core/Emulator/Devices/Video/VgaIoPortHandler.cs

```diff
@@ -0,0 +1,354 @@
+using Spice86.Core.Emulator.VM;
+using Spice86.Shared.Interfaces;
+
+/// <summary>
+///     I/O port handler for the VGA card.
+/// </summary>
+public class VgaIoPortHandler : DefaultIOPortHandler {
+    private readonly AttributeControllerRegisters _attributeRegisters;
+    private readonly CrtControllerRegisters _crtRegisters;
+    private readonly DacRegisters _dacRegisters;
+    private readonly GeneralRegisters _generalRegisters;
+    private readonly GraphicsControllerRegisters _graphicsRegisters;
+    private readonly ILoggerService _logger;
+    private readonly SequencerRegisters _sequencerRegisters;
+    private bool _attributeDataMode;
+
+    /// <summary>
+    ///     Create a new VGA I/O port handler.
+    /// </summary>
+    /// <param name="machine"></param>
+    /// <param name="loggerService"></param>
+    /// <param name="configuration"></param>
+    /// <param name="videoState"></param>
+    public VgaIoPortHandler(Machine machine, ILoggerService loggerService, Configuration configuration, IVideoState videoState) :
+        base(machine, configuration, loggerService) {
+        _logger = loggerService.WithLogLevel(LogEventLevel.Information);
+
+        // Initialize registers.
+        _attributeRegisters = videoState.AttributeControllerRegisters;
+        _graphicsRegisters = videoState.GraphicsControllerRegisters;
+        _sequencerRegisters = videoState.SequencerRegisters;
+        _crtRegisters = videoState.CrtControllerRegisters;
+        _generalRegisters = videoState.GeneralRegisters;
+        _dacRegisters = videoState.DacRegisters;
+    }
+
+    private static IEnumerable<int> HandledPorts =>
+        new SortedSet<int> {
+            Ports.AttributeAddress,
+            Ports.AttributeData,
+            Ports.CrtControllerAddress,
+            Ports.CrtControllerAddressAlt,
+            Ports.CrtControllerAddressAltMirror1,
+            Ports.CrtControllerAddressAltMirror2,
+            Ports.CrtControllerData,
+            Ports.CrtControllerDataAlt,
+            Ports.CrtControllerDataAltMirror1,
+            Ports.CrtControllerDataAltMirror2,
+            Ports.DacAddressReadIndex,
+            Ports.DacAddressWriteIndex,
+            Ports.DacData,
+            Ports.DacPelMask,
+            Ports.DacStateRead,
+            Ports.FeatureControlRead,
+            Ports.FeatureControlWrite,
+            Ports.FeatureControlWriteAlt,
+            Ports.GraphicsControllerAddress,
+            Ports.GraphicsControllerData,
+            Ports.InputStatus0Read,
+            Ports.InputStatus1Read,
+            Ports.InputStatus1ReadAlt,
+            Ports.MiscOutputRead,
+            Ports.MiscOutputWrite,
+            Ports.SequencerAddress,
+            Ports.SequencerData,
+            Ports.CgaModeControl,
+            Ports.CgaColorSelect
+        };
+
+    /// <inheritdoc />
+    public override void InitPortHandlers(IOPortDispatcher ioPortDispatcher) {
+        foreach (int port in HandledPorts) {
```

Comment:
Would be nice to have IOPortDispatcher supporting adding a list, could be used by other devices (most?)

Reply:
Could be, I'm not sure it belongs in this PR or even if it makes any difference in performance or ease-of-development.